### PR TITLE
Save daily docket changes to Redux

### PR DESCRIPTION
**Pull request by slj**
Connects Connects #2473

### CODE REVIEW

File: app/assets/stylesheets/_hearings.scss [hunk 1]
```diff
@@ -102,10 +102,9 @@ $color-hearings-saving: #777;
         }
       }
 
-      // TODO: get rule into a SaveData component
       .saving {
-        font-style: italic;
-        color: $color-hearings-saving;
+        width: 75%;
+        text-align: right;
```
Comment (slj) [inline app/assets/stylesheets/_hearings.scss]: There's another rule that sets the font style and color for this class in general. When this element appears as part of a table header, the rules above place it correctly.

File: app/assets/stylesheets/_hearings.scss [hunk 2]
```diff
@@ -178,7 +178,7 @@ $color-hearings-saving: #777;
     }
   }
 
-  .transcriptRequired {
+  .transcriptRequested {
```
Comment (slj) [inline app/assets/stylesheets/_hearings.scss]: "Required" changed to "Requested" for a bit more consistency

File: app/assets/stylesheets/_hearings.scss [hunk 3]
```diff
@@ -219,9 +219,15 @@ $color-hearings-saving: #777;
...
-  overflow: hidden;
   margin-bottom: 1.765em;
 
+  &::after {
+    content: ' ';
+    display: block;
+    line-height: 0;
+    clear: both;
+  }
+
```
Comment (slj) [inline app/assets/stylesheets/_hearings.scss]: `overflow: hidden` often works but in this case it was enabling scrolling, probably due to some padding in the elements. They key here is `line-height: 0` to keep the break from taking any vertical space.

File: app/assets/stylesheets/_hearings.scss [hunk 4]
```diff
@@ -318,3 +330,22 @@ $color-hearings-saving: #777;
...
+    padding: 0;
+    width: 1em;
+    height: 1em;
+    vertical-align: middle;
+  }
+
+  .cf-loading-button-text {
+    display: none;
+  }
+}
```
Comment (slj) [inline app/assets/stylesheets/_hearings.scss]: This is general styling for the element that shows "Saving... " with a spinning logo.

File: app/controllers/hearings_controller.rb [hunk 5]
```diff
@@ -15,6 +15,15 @@ def logo_path
...
 
+  # Temp action to simulate the autosave feature
+  def save_data
+    sleep(1.0 * 0.4)
+    Rails.logger.info "**************************************************"
+    Rails.logger.info "Saving #{params}"
+    Rails.logger.info "**************************************************"
+    render status: 200, json: {}
+  end
+
```
Comment (slj) [inline app/controllers/hearings_controller.rb]: This is temporary but will be helpful for the QA team to validate that a save request is made - this PR does not do any actual saving to the database.

File: client/app/components/Checkbox.jsx [hunk 6]
```diff
@@ -35,6 +36,7 @@ export default class Checkbox extends React.Component {
           type="checkbox"
           id={name}
           checked={value}
+          tabIndex={tabIndex}
```
Comment (slj) [inline client/app/components/Checkbox.jsx]: Adding a tab index creates a better UX on the daily do

File: client/app/hearings/AutosavePrompt.jsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  saving: state.saving
+});
+
+export default connect(
+  mapStateToProps
+)(AutosavePrompt);
+
+AutosavePrompt.propTypes = {
+  saving: PropTypes.bool
+};
```
Comment (slj) [inline client/app/hearings/AutosavePrompt.jsx]: Simple component that takes a property saving, which if true, renders "Saving..." with a spinner, or "Last saved at ".

File: client/app/hearings/CheckboxContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(CheckboxContainer);
+
+CheckboxContainer.propTypes = {
+  id: PropTypes.string.isRequired,
+  label: PropTypes.string.isRequired
+};
```
Comment (slj) [inline client/app/hearings/CheckboxContainer.jsx]: By wrapping the `Transcript Required` checkbox with a container, it's easy to map the value property the redux store without editing the actual Checkbox component.

File: client/app/hearings/DailyDocket.jsx [hunk 9]
```diff
@@ -1,26 +1,28 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
-import SearchableDropdown from '../components/SearchableDropdown';
-import Checkbox from '../components/Checkbox';
+import TextareaContainer from './TextareaContainer';
+import DropdownContainer from './DropdownContainer';
+import CheckboxContainer from './CheckboxContainer';
```
Comment (slj) [inline client/app/hearings/DailyDocket.jsx]: Again, these wrappers enable us to easily map value properties to the redux store without tweaking the core components.

File: client/app/hearings/DailyDocket.jsx [hunk 10]
```diff
@@ -36,30 +38,9 @@ const getDate = (date, timezone) => {
...
-      let text = document.getElementsByClassName('saving')[0].textContent;
-
-      if (text.startsWith('Saving...')) {
-        document.getElementsByClassName('saving')[0].innerHTML = `Last saved at ${now()}`;
-      } else {
-        document.getElementsByClassName('saving')[0].innerHTML = 'Saving...';
-      }
-    }, 3000);
-  }
-
```
Comment (slj) [inline client/app/hearings/DailyDocket.jsx]: Removed these blocks of code because the AutosavePrompt component handles this now - no more alternating "Saving..."/"Last Saved..." automatically.

File: client/app/hearings/DailyDocket.jsx [hunk 11]
```diff
@@ -103,53 +84,54 @@ export class DailyDocket extends React.Component {
...
-                    name={`transcript_requested_${index}`}
-                    onChange={() => {
-                      return true;
-                    }}
-                    value={false}
-                  ></Checkbox>
+                    defaultValue={hearing.transcriptRequested}
+                    action="updateDailyDocketTranscript"
+                    tabIndex={`${(index * 10) + 5}`}
+                  ></CheckboxContainer>
```
Comment (slj) [inline client/app/hearings/DailyDocket.jsx]: The general approach with Textarea/Dropdown/CheckboxContainer was to create value properties on the redux store in the format `<resourceId>.<property>`.  This way when it's time to for an ApiUtil request, a quick parse of the property easily generates the payload. So, for example, if hearing properties `1.notes`, `1.disposition`, and `1.hold_open` have values "reschedule"/held/60 respectively, a quick parse will yield a payload of `{id:1, model: {notes: "reschedule", disposition: "held", hold_open: 60}}`

File: client/app/hearings/DailyDocket.jsx [hunk 12]
```diff
@@ -158,7 +140,7 @@ export class DailyDocket extends React.Component {
           )}
         </table>
       </div>
-      <div className="cf-alt--actions cf-alt--app-width">
+      <div className="cf-actions cf-app-width">
```
Comment (slj) [inline client/app/hearings/DailyDocket.jsx]: No more application ALT layout.

File: client/app/hearings/DailyDocketContainer.jsx [hunk 13]
```diff
@@ -1,17 +1,18 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
-import * as Actions from './actions/Dockets';
+import { handleServerError } from './actions/hearings';
+import { populateDockets } from './actions/dockets';
```
Comment (slj) [inline client/app/hearings/DailyDocketContainer.jsx]: Separating general Hearings actions from those pertaining to Dockets.

File: client/app/hearings/DropdownContainer.jsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  mapDispatchToProps
+)(DropdownContainer);
+
+DropdownContainer.propTypes = {
+  label: PropTypes.string,
+  name: PropTypes.string.isRequired,
+  onChange: PropTypes.func,
+  options: PropTypes.array,
+  tabIndex: PropTypes.string
+};
```
Comment (slj) [inline client/app/hearings/DropdownContainer.jsx]: Again wrapping core component with container to easily map value properties to the redux store.

File: client/app/hearings/HearingPrepContainer.jsx [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+});
+
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(HearingPrepContainer);
+
+HearingPrepContainer.propTypes = {
+  hearings: PropTypes.object
+};
```
Comment (slj) [inline client/app/hearings/HearingPrepContainer.jsx]: This is main container for Hearing Prep, factored out of `hearings/index.jsx`. This makes is easier to respond to the `onbeforeunload` event when closing the window.

File: client/app/hearings/HearingWorksheetContainer.jsx [hunk 16]
```diff
@@ -1,7 +1,8 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
-import * as Actions from './actions/Dockets';
+import { handleServerError } from './actions/hearings';
+import { populateWorksheet } from './actions/worksheet';
```
Comment (slj) [inline client/app/hearings/HearingWorksheetContainer.jsx]: Again, due to separation of actions.

File: client/app/hearings/TextareaContainer.jsx [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+});
+
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(TextareaContainer);
+
+TextareaContainer.propTypes = {
+  id: PropTypes.string.isRequired
+};
```
Comment (slj) [inline client/app/hearings/TextareaContainer.jsx]: Again, wrapping core component with container to easily map value properties to the redux store.

File: client/app/hearings/actions/Dockets.js [hunk 18]
```diff
@@ -1,22 +0,0 @@
...
-    worksheet
-  }
-});
-
-export const handleServerError = (err) => ({
-  type: Constants.HANDLE_SERVER_ERROR,
-  payload: {
-    err
-  }
-});
```
Comment (slj) [inline client/app/hearings/actions/Dockets.js]: Replaced by `dockets.js` - I prefer component filenames to be capitalized and action filenames to stay in lowercase.

File: client/app/hearings/actions/dockets.js [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+        type: 'TOGGLE_SAVING'
+      }),
+      (err) => {
+        dispatch(handleServerError(err));
+      }
+    );
+};
+
+export const updateDailyDocketNotes = (prop, value) => {
+  return updatedState({
```
Comment (slj) [inline client/app/hearings/actions/dockets.js]: Most action creators will call `updatedState()`

File: client/app/hearings/constants/constants.js [hunk 20]
```diff
@@ -1,5 +1,10 @@
 // actions
 //
+export const HANDLE_SERVER_ERROR = 'HANDLE_SERVER_ERROR';
+
```
Comment (slj) [inline client/app/hearings/constants/constants.js]: Just moved this to the top

File: client/app/hearings/reducers/index.js [hunk 21]
```diff
@@ -5,25 +5,48 @@
...
+      saving: { $set: !state.saving }
+    });
+
   case 'HANDLE_SERVER_ERROR':
-    return Object.assign({}, state, {
-      serverError: action.payload.err
+    return update(state, {
+      serverError: { $set: action.payload.err }
     });
+
```
Comment (slj) [inline client/app/hearings/reducers/index.js]: Using `immutability-helper` instead of `Object.assign`

File: client/app/hearings/utils.js [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+  Object.keys(data).forEach((key) => {
+    saveModel[key.split('.')[1]] = data[key];
+    if (saveModelId === null) {
+      saveModelId = key.split('.')[0];
+    }
+  });
+
+  return { id: saveModelId,
+    model: saveModel };
+};
```
Comment (slj) [inline client/app/hearings/utils.js]: This is a utility for what was mentioned [earlier](https://github.com/department-of-veterans-affairs/caseflow/pull/2809/commits/2672a069d9d8e904489bd45ea3e0cb0f2267b0a5#r130888633).

**Comment by NickHeiner** [comment]
Hi. This is a pretty big PR, and it looks like some of the stuff is not related to #2473. I'm going to go through and comment on the specific sections that look extraneous to me. Perhaps I am confused about that ticket and you can help me understand how the changes are all related.

File: app/assets/stylesheets/_hearings.scss [hunk 23]
```diff
@@ -102,10 +102,9 @@ $color-hearings-saving: #777;
         }
       }
 
-      // TODO: get rule into a SaveData component
```
Comment (NickHeiner) [inline app/assets/stylesheets/_hearings.scss]: Why are there styling changes in this PR? I don't see anything about styling in #2473.

File: app/controllers/hearings_controller.rb [hunk 24]
```diff
@@ -15,6 +15,15 @@ def logo_path
     hearings_dockets_path
   end
 
+  # Temp action to simulate the autosave feature
```
Comment (NickHeiner) [inline app/controllers/hearings_controller.rb]: Why is ruby code included in this PR? The ticket specifically says:

> No backend integration needed in this ticket

File: client/app/components/Checkbox.jsx (hunk 6 above)
Comment (NickHeiner) [inline client/app/components/Checkbox.jsx]: A better UX is nice, but this doesn't seem like it's related to #2473. Let's split this out into a separate ticket. 

Did the designer ask for this change?

File: client/app/hearings/AutosavePrompt.jsx [hunk 25]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
```
Comment (NickHeiner) [inline client/app/hearings/AutosavePrompt.jsx]: Why is this here? I don't see anything about autosaving in #2473.

**Comment by slj** [comment]
While everything may not be precisely related to #2473, everything there is closely related to it has QA validation in mind, so please consider reviewing with more of an effort to understand than to point out what seems out of place.

File: client/app/hearings/DailyDocket.jsx (hunk 12 above)
Comment (NickHeiner) [inline client/app/hearings/DailyDocket.jsx]: How is this related to #2473?

File: client/app/hearings/DailyDocketContainer.jsx (hunk 13 above)
Comment (NickHeiner) [inline client/app/hearings/DailyDocketContainer.jsx]: This sounds like a good refactor, but this PR is already quite large. If removing the other extraneous stuff doesn't get us down to around `200` additions, then it would be good to do this in a separate PR as well.

File: client/app/hearings/HearingPrepContainer.jsx [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+import { prepSaveData } from './utils';
+import ApiUtil from '../util/ApiUtil';
+import DocketsContainer from './DocketsContainer';
+import DailyDocketContainer from './DailyDocketContainer';
+import HearingWorksheetContainer from './HearingWorksheetContainer';
+
+export const saveBeforeWindowCloses = () => (dispatch, getState) => {
+  const dataToSave = prepSaveData(getState().save);
+
+  ApiUtil.post(
```
Comment (NickHeiner) [inline client/app/hearings/HearingPrepContainer.jsx]: Why are we making a backend call? #2473 specifically says:

> No backend integration needed in this ticket

File: client/app/hearings/HearingPrepContainer.jsx [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+    '/hearings/save_data', { data: dataToSave }
+  ).then(
+    // w/o then(), POST request seems to fail
+  );
+};
+
+export class HearingPrepContainer extends React.PureComponent {
+
+  componentDidMount = () => {
+    window.onbeforeunload = () => {
```
Comment (NickHeiner) [inline client/app/hearings/HearingPrepContainer.jsx]: This seems like it belongs in another ticket – maybe #2475?

File: client/app/hearings/actions/Dockets.js (hunk 18 above)
Comment (NickHeiner) [inline client/app/hearings/actions/Dockets.js]: I'm fine with whatever naming convention we choose, but if this PR is already too big, then refactors like this should be split out separately.

File: client/app/hearings/actions/dockets.js [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+import diff from 'object-diff';
+
+export const updateUtil = {
+  lastChangeTimestamp: null,
+  lastSave: {},
+  diffObjects: diff
+};
+
+/* eslint-disable no-unused-vars */
+/* eslint-disable no-eval */
```
Comment (NickHeiner) [inline client/app/hearings/actions/dockets.js]: This rule should not be disabled. It is extremely unlikely that we actually need to use `eval`. If you need help figuring out how to do something without using `eval`, @sharonwarner or @joofsh or myself can pair with you.

File: client/app/hearings/actions/dockets.js [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+/* eslint-disable no-console */
+export const updatedState = (action, milliseconds = 1000) =>
+  (dispatch, getState) => {
+
+    dispatch(action);
+
+    updateUtil.lastChangeTimestamp = Number(new Date());
+
+    // create a function that, after a moment (default 1000ms),
```
Comment (NickHeiner) [inline client/app/hearings/actions/dockets.js]: If you want a function that only runs so often, check out `_.debounce`.

File: client/app/hearings/actions/dockets.js [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+          updateUtil.lastSave = getState().save;
+        }
+      }, ${milliseconds});`
+    );
+
+  };
+/* eslint-enable no-unused-vars */
+/* eslint-enable no-eval */
+
+export const saveData = (data, dispatch, url = '/hearings/save_data') => {
```
Comment (NickHeiner) [inline client/app/hearings/actions/dockets.js]: This action is touching the backend, which is out of scope for #2473.

File: client/app/hearings/reducers/index.js [hunk 31]
```diff
@@ -5,25 +5,48 @@
...
+import update from 'immutability-helper';
 
 export const mapDataToInitialState = function(state = {}) {
   return state;
 };
 
+export const newState = (action, state) => {
+  return update(state, {
+    [action.payload.prop]: { $set: action.payload.value },
+    save: { $apply: (_save) =>
```
Comment (NickHeiner) [inline client/app/hearings/reducers/index.js]: Is this for backend integration? If so, we don't need it for #2473.

File: client/app/hearings/reducers/index.js [hunk 32]
```diff
@@ -5,25 +5,48 @@
...
+      dockets: { $set: action.payload.dockets }
     });
+
+  case 'UPDATE_DAILY_DOCKET_NOTES':
+  case 'UPDATE_DAILY_DOCKET_TRANSCRIPT':
+  case 'UPDATE_DAILY_DOCKET_ACTION':
+    return newState(action, state);
+
   case 'POPULATE_WORKSHEET':
-    return Object.assign({}, state, {
```
Comment (NickHeiner) [inline client/app/hearings/reducers/index.js]: Please do not migrate existing reducers to use `immutability-helper`. I specifically asked that we leave that for a future PR in my earlier comment:

image

File: config/routes.rb [hunk 33]
```diff
@@ -72,6 +72,8 @@
     resources :worksheets, only: [:update, :show]
   end
 
+  post 'hearings/save_data' => "hearings#save_data"
```
Comment (NickHeiner) [inline config/routes.rb]: This should not be necessary, since #2473 does not require backend integration.

**Comment by NickHeiner** [comment]
> While everything may not be precisely related to #2473, everything there is closely related to it has QA validation in mind, so please consider reviewing with more of an effort to understand than to point out what seems out of place.

Sorry, I didn't see this until I finished my review. Please take a look at my comments, and I'm happy to chat in person if there are questions.
